pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +147,290 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+			return pkgerror.Wrapf(err, "failed to fence rbd volumes")
+		}
+	}
+
+	return nil
+}
+
+func getCephVolumesInUse(cluster *cephv1.CephCluster, volumesInUse []v1.UniqueVolumeName) []string {
+	var rbdVolumesInUse []string
+
+	for _, volume := range volumesInUse {
+		volumesInuseRemoveK8sPrefix := strings.TrimPrefix(string(volume), "kubernetes.io/csi/")
+		splitVolumeInUseBased := strings.Split(volumesInuseRemoveK8sPrefix, "^")
+		logger.Infof("volumeInUse after split based on '^' %v", splitVolumeInUseBased)
+
+		if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == fmt.Sprintf("%s.rbd.csi.ceph.com", cluster.Namespace) {
+			rbdVolumesInUse = append(rbdVolumesInUse, splitVolumeInUseBased[1])
+		}
+	}
+	return rbdVolumesInUse
+}
+
+func listRBDPV(listPVs *corev1.PersistentVolumeList, cluster *cephv1.CephCluster, rbdVolumesInUse []string) []corev1.PersistentVolume {
+	var listRbdPV []corev1.PersistentVolume
+
+	for _, pv := range listPVs.Items {
+		if pv.Spec.CSI.Driver == fmt.Sprintf("%s.rbd.csi.ceph.com", cluster.Namespace) {
+			for _, rbdVolume := range rbdVolumesInUse {
+				if pv.Spec.CSI.VolumeHandle == rbdVolume {
+					listRbdPV = append(listRbdPV, pv)
+				}
+			}
+		}
+	}
+	return listRbdPV
+}
+
+func (c *clientCluster) fenceRbdImage(
+	ctx context.Context, node *corev1.Node, cluster *cephv1.CephCluster,
+	clusterInfo *cephclient.ClusterInfo, rbdPV corev1.PersistentVolume) error {
+
+	logger.Debugf("rbd PV NAME %v", rbdPV.Spec.CSI.VolumeAttributes)
+	args := []string{"status", fmt.Sprintf("%s/%s", rbdPV.Spec.CSI.VolumeAttributes["pool"], rbdPV.Spec.CSI.VolumeAttributes["imageName"])}
+	cmd := cephclient.NewRBDCommand(c.context, clusterInfo, args)
+	cmd.JsonOutput = true
+
+	buf, err := cmd.Run()
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to list watchers for pool/imageName %s/%s.", rbdPV.Spec.CSI.VolumeAttributes["pool"], rbdPV.Spec.CSI.VolumeAttributes["imageName"])
+	}
+
+	logger.Infof("rbd status to get the client ips %v", string(buf))
+	ips, err := rbdStatusUnMarshal(buf)
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to unmarshal rbd status output")
+	}
+	if len(ips) != 0 {
+		err = c.createNetworkFence(ctx, rbdPV, node, cluster, ips)
+		if err != nil {
+			return pkgerror.Wrapf(err, "failed to create network fence for node %q", node.Name)
+		}
+	}
+
+	return nil
+}
+
+func rbdStatusUnMarshal(output []byte) ([]string, error) {
+	type rbdStatus struct {
+		Watchers []struct {
+			Address string `json:"address"`
+		} `json:"watchers"`
+	}
+
+	var rbdStatusObj rbdStatus
+	err := json.Unmarshal([]byte(output), &rbdStatusObj)
+	if err != nil {
+		return []string{}, pkgerror.Wrapf(err, "failed to unmarshal rbd status output")
+	}
+
+	watcherIPlist := []string{}
+	for _, watcher := range rbdStatusObj.Watchers {
+		// split with separation '/' to remove nounce and concatenating `/32` to define a network with only one IP address
+		address := strings.Split(watcher.Address, "/")[0] + "/32"
```

Comment:
Is there a unit test for this parsing?

Reply:
added for this and for couple of other cases too